hazelcast-client/src/test/java/com/hazelcast/client/test/TestAwareClientFactory.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.test;
+
+import static com.hazelcast.test.AbstractHazelcastClassRunner.getTestMethodName;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.hazelcast.client.HazelcastClient;
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.test.TestAwareInstanceFactory;
+
+/**
+ * Per test-method factory for Hazelcast clients (and also members as it inherits from {@link TestAwareInstanceFactory}).
+ * It configures new clients in the same way as it's done for members in {@link TestAwareInstanceFactory#newHazelcastInstance(com.hazelcast.config.Config)}.
+ */
+public class TestAwareClientFactory extends TestAwareInstanceFactory {
+    protected final Map<String, List<HazelcastInstance>> perMethodClients = new ConcurrentHashMap<String, List<HazelcastInstance>>();
+
+    /**
+     * Creates new client instance which uses in its network configuration the first member created by this factory. The value
+     * {@link com.hazelcast.test.AbstractHazelcastClassRunner#getTestMethodName()} is used as a cluster group name.
+     */
+    public HazelcastInstance newHazelcastClient(ClientConfig config) {
+        if (config == null) {
+            config = new ClientConfig();
+        }
+        config.getGroupConfig().setName(getTestMethodName());
+        List<HazelcastInstance> members = getOrInitInstances(perMethodMembers);
+        if (members.isEmpty()) {
+            throw new IllegalStateException("Members have to be created first");
+        }
+        config.getNetworkConfig().addAddress("127.0.0.1:" + getPort(members.get(0)));
```

Comment:
Would you mind justifying why it's OK to take always the first member? Why don't we care about the (possible) others? I mean comment it here, I just want to be sure.

Reply:
It's a very good question. ;)
The answer is, there is no reason to fill only the first member. It was sufficient for the original purpose (inside the `TlsFunctionalTest`), but it will be more natural to fill all existing members.
I'll fix it.

Reply:
Ok, in that case please fix in on both places. I guess a simple random choice of members to fill is enough.